In [1]:
import keras
keras.__version__

Using TensorFlow backend.


'2.2.4'

In [3]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

In [4]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


1. Conv2D(filters, kernel_size, strides=(1, 1), padding='valid', data_format=None, dilation_rate=(1, 1), activation=None, use_bias=True, kernel_initializer='glorot_uniform', bias_initializer='zeros', kernel_regularizer=None, bias_regularizer=None, activity_regularizer=None, kernel_constraint=None, bias_constraint=None)

    - Conv1D와 동일하지만, 다른 점은 1차원이 아닌 2차원이라는 점이다.
    - 28 * 28 사이즈이므로 input_shape은 (28, 28, 1)이 된다.
    

2. MaxPooling2D(pool_size=(2, 2), strides=None, padding='valid', data_format=None)

    - __pool_size__: 가로, 세로 축소 비율을 결정한다. 여기서는 (2, 2)로 절반으로 축소된다.
    - __model.summary()__를 보게 되면, 26 -> 13, 11 -> 5를 보게 되면 알 수 있다.

In [5]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

In [6]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                36928     
__________

__Flatten__: 이제 (3, 3, 64)의 형태에서 1차원의 형태인 (576)으로 변경해주는 작업을 진행한다.


이후 64개로 나눈 다음 10개 (0 ~ 9)로 변경해주는 작업을 한다.


In [7]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [9]:
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 32s 528us/step - loss: 0.0873 - acc: 0.9736
Epoch 2/5
60000/60000 [==============================] - 32s 532us/step - loss: 0.0438 - acc: 0.9868
Epoch 3/5
60000/60000 [==============================] - 32s 541us/step - loss: 0.0313 - acc: 0.9904
Epoch 4/5
60000/60000 [==============================] - 32s 529us/step - loss: 0.0239 - acc: 0.9926
Epoch 5/5
60000/60000 [==============================] - 31s 515us/step - loss: 0.0189 - acc: 0.99410s - loss: 0.0189 - acc: 0.99


컴파일 이후 학습하는 과정이다.


__rmsprop__: 과거의 모든 기울기를 균일하게 더하지 않고, 새로운 기울기의 정보만 반영하도록 한다.


__categorical_crossentropy__: 클래스가 3개 이상일 경우 사용 


__accuracy__:  다중클래스분류 문제에서 평가기준을 ‘accuracy’로 지정했을 경우 내부적으로 categorical_accuracy() 함수를 이용하여 정확도가 계산.

In [10]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 1s 145us/step


In [11]:
test_acc

0.9897

In [12]:
test_loss

0.031904928047856086

## 결론

- __정확도: 98.97%__